### ISSUE TITLE
Waiting For Edge Crashes

### ISSUE THREAD

**Issue by foohyfooh** [esp-rs/esp-idf-hal#254]
I have a button setup like the image below
![Schematic](https://github.com/esp-rs/esp-idf-hal/assets/3344332/ff2432f8-31a4-44c4-a03d-9dc12980fc8f)

and the project is creating using the esp-idf-template given by `cargo generate --git https://github.com/esp-rs/esp-idf-template cargo` with the code as follows

```rs
use esp_idf_sys as _; // If using the `binstart` feature of `esp-idf-sys`, always keep this module imported
use esp_idf_hal::{gpio::*, peripherals::Peripherals};
use futures::executor;

fn main() {
    // It is necessary to call this function once. Otherwise some patches to the runtime
    // implemented by esp-idf-sys might not link properly. See https://github.com/esp-rs/esp-idf-template/issues/71
    esp_idf_sys::link_patches();
    // Bind the log crate to the ESP Logging facilities
    esp_idf_svc::log::EspLogger::initialize_default();

    let peripherals = Peripherals::take().unwrap();
    let mut button = PinDriver::input_output(peripherals.pins.gpio33).unwrap();
    button.set_pull(Pull::Up).unwrap();
    button.enable_interrupt().unwrap();
    button.set_interrupt_type(InterruptType::AnyEdge).unwrap();

    loop {
        _ = executor::block_on(button.wait_for_any_edge());
        log::info!("Got an edge");
    }
}
```
but pressing the button crashes with the output in the [logs](https://github.com/esp-rs/esp-idf-hal/files/11556728/logs.txt)

**Comment by foohyfooh** [comment]
Use the subscribe method with the following code doesn't appear to trigger the callback and if I uncomment the line setting the interrupt type it produces a panic like the above code shown in the [logs](https://github.com/esp-rs/esp-idf-hal/files/11566860/logs2.txt)

```rs
use esp_idf_sys as _; // If using the `binstart` feature of `esp-idf-sys`, always keep this module imported
use esp_idf_hal::{gpio::*, peripherals::Peripherals};
use futures::executor;
use std::{thread, time::Duration};

const USE_INTERRUPT: bool = false;
const PULLUP: bool = true;

fn subscription_callback() {
    log::info!("Got subscription");
}

fn main() {
    // It is necessary to call this function once. Otherwise some patches to the runtime
    // implemented by esp-idf-sys might not link properly. See https://github.com/esp-rs/esp-idf-template/issues/71
    esp_idf_sys::link_patches();
    // Bind the log crate to the ESP Logging facilities
    esp_idf_svc::log::EspLogger::initialize_default();

    log::info!("Using {}", if PULLUP {"Pull Up"} else {"Pull Down"});
    log::info!("Using {}", if USE_INTERRUPT {"Interrupt"} else {"Subscribe"});

    let peripherals = Peripherals::take().unwrap();
    let mut button = PinDriver::input_output(peripherals.pins.gpio33).unwrap();
    button.set_pull(if PULLUP {Pull::Up} else {Pull::Down}).unwrap();

    if USE_INTERRUPT {
        button.enable_interrupt().unwrap();
        button.set_interrupt_type(InterruptType::AnyEdge).unwrap();
    } else {
        unsafe {
            button.subscribe(subscription_callback).unwrap();
        }
        // button.set_interrupt_type(InterruptType::AnyEdge).unwrap();
    }

    loop {
        if USE_INTERRUPT {
            _ = executor::block_on(button.wait_for_any_edge());
            log::info!("Got an edge");
        } else {
            thread::sleep(Duration::from_millis(10));
        }
    }
}
```

**Comment by Vollbrecht** [comment]
The subscription callback cannot run any complicated function. The info crate and all println function cannot be used inside an ISR. the callback is run in the context of the ISR. you can try sending an simple state variable out of the ISR or use some other ISR save method that provides signaling and act on it. Such variables needs to be protected by a CriticalSection etc

**Comment by foohyfooh** [comment]
Okay. Use a mutable static variable in the callback works and the loop works. That was my bad for using `log::info!` in the callback.

**Comment by ivmarkov** [comment]
@foohyfooh  I don't think the **root cause** why the async stuff does not work was ever addressed, was it?

TL;DR:
It does not work, because you are using the stock executor from the `futures` crate. You should be using [this one](https://github.com/ivmarkov/edge-executor) instead, i.e. enable the `edge-executor` feature of the `esp-idf-hal` crate.

The long(ish) story is that standard executors are not compatible with stuff which calls their `Waker`s directly from an ISR context. My customization of the `async-task` executor from above is compatible though.

Let me know if it works for you with `edge-executor` so I can close this.

**Comment by foohyfooh** [comment]
`edge-executor` is giving the desired outcome. Thanks for the help.
